jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientTLSTest.java

```diff
@@ -265,51 +264,6 @@ public void handshakeFailed(Event event, Throwable failure)
         assertTrue(clientLatch.await(1, TimeUnit.SECONDS));
     }
 
-    // In JDK 11+, a mismatch on the client does not generate any bytes towards
-    // the server, while in previous JDKs the client sends to the server the close_notify.
-    // @EnabledOnJre({JRE.JAVA_8, JRE.JAVA_9, JRE.JAVA_10})
-    @Disabled("No longer viable, TLS protocol behavior changed in 8u272")
-    public void testMismatchBetweenTLSProtocolAndTLSCiphersOnClient() throws Exception
-    {
-        SslContextFactory serverTLSFactory = createServerSslContextFactory();
-        startServer(serverTLSFactory, new EmptyServerHandler());
-
-        CountDownLatch serverLatch = new CountDownLatch(1);
-        connector.addBean(new SslHandshakeListener()
-        {
-            @Override
-            public void handshakeFailed(Event event, Throwable failure)
-            {
-                serverLatch.countDown();
-            }
-        });
-
-        SslContextFactory clientTLSFactory = createClientSslContextFactory();
-        // TLS 1.1 protocol, but only TLS 1.2 ciphers.
-        clientTLSFactory.setIncludeProtocols("TLSv1.1");
-        clientTLSFactory.setIncludeCipherSuites("TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256");
-        startClient(clientTLSFactory);
-
-        CountDownLatch clientLatch = new CountDownLatch(1);
-        client.addBean(new SslHandshakeListener()
-        {
-            @Override
-            public void handshakeFailed(Event event, Throwable failure)
-            {
-                clientLatch.countDown();
-            }
-        });
-
-        assertThrows(ExecutionException.class, () ->
-            client.newRequest("localhost", connector.getLocalPort())
-                .scheme(HttpScheme.HTTPS.asString())
-                .timeout(5, TimeUnit.SECONDS)
-                .send());
-
-        assertTrue(serverLatch.await(1, TimeUnit.SECONDS));
-        assertTrue(clientLatch.await(1, TimeUnit.SECONDS));
-    }
-
```

Comment:
I think we need to rewrite this test, as we want to know what happens in case of mismatch.
Since 8u272 introduced TLS 1.3 and it's now hopefully more similar to Java 11, I think that this test could be rewritten with the Java 11 behavior.

Reply:
On Jetty 9.4.x that would mean developers using older versions of Java 8 would then trigger this issue.

If you feel strongly about this, then we should probably focus this test rewrite on Jetty 10+

Reply:
I think the test should be in 9.4.x too, as it just popped up in sponsored support.

Reply:
I don't think this test case is valid anymore.

If you attempt to create a mismatch that is mentioned in the old testcase ..

``` java
    // In JDK 11+, a mismatch on the client does not generate any bytes towards
    // the server, while in previous JDKs the client sends to the server the close_notify.
    @EnabledOnJre({JRE.JAVA_8, JRE.JAVA_9, JRE.JAVA_10})
    public void testMismatchBetweenTLSProtocolAndTLSCiphersOnClient() throws Exception
```

You get the same behavior as in Java 11+

This behavior started in Java 8u272.

```
javax.net.ssl.SSLHandshakeException: No appropriate protocol (protocol is disabled or cipher suites are inappropriate)
	at org.eclipse.jetty.client.util.FutureResponseListener.getResult(FutureResponseListener.java:118)
	at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:101)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:730)
	at org.eclipse.jetty.client.HttpClientTLSTest.lambda$testMismatchBetweenTLSProtocolAndTLSCiphersOnClient$3(HttpClientTLSTest.java:307)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:55)
	at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:37)
	at org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3007)
	at org.eclipse.jetty.client.HttpClientTLSTest.testMismatchBetweenTLSProtocolAndTLSCiphersOnClient(HttpClientTLSTest.java:303)
Caused by: javax.net.ssl.SSLHandshakeException: No appropriate protocol (protocol is disabled or cipher suites are inappropriate)
	at sun.security.ssl.HandshakeContext.<init>(HandshakeContext.java:171)
	at sun.security.ssl.ClientHandshakeContext.<init>(ClientHandshakeContext.java:98)
	at sun.security.ssl.TransportContext.kickstart(TransportContext.java:220)
	at sun.security.ssl.SSLEngineImpl.writeRecord(SSLEngineImpl.java:159)
	at sun.security.ssl.SSLEngineImpl.wrap(SSLEngineImpl.java:130)
	at sun.security.ssl.SSLEngineImpl.wrap(SSLEngineImpl.java:110)
	at javax.net.ssl.SSLEngine.wrap(SSLEngine.java:511)
	at org.eclipse.jetty.io.ssl.SslConnection.wrap(SslConnection.java:424)
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.flush(SslConnection.java:1089)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:277)
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:381)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.process(HttpSenderOverHTTP.java:258)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:223)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(HttpSenderOverHTTP.java:65)
	at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:212)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:84)
	at org.eclipse.jetty.client.HttpChannel.send(HttpChannel.java:125)
	at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:241)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:269)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:125)
	at org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(HttpDestinationOverHTTP.java:38)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:377)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:332)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:311)
	at org.eclipse.jetty.client.HttpDestination.succeeded(HttpDestination.java:246)
	at org.eclipse.jetty.client.AbstractConnectionPool.proceed(AbstractConnectionPool.java:313)
	at org.eclipse.jetty.client.AbstractConnectionPool$FutureConnection.succeeded(AbstractConnectionPool.java:533)
	at org.eclipse.jetty.client.AbstractConnectionPool$FutureConnection.succeeded(AbstractConnectionPool.java:511)
	at org.eclipse.jetty.util.Promise$Wrapper.succeeded(Promise.java:130)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onOpen(HttpConnectionOverHTTP.java:133)
	at org.eclipse.jetty.io.ssl.SslConnection.onOpen(SslConnection.java:372)
	at org.eclipse.jetty.io.SelectorManager.connectionOpened(SelectorManager.java:324)
	at org.eclipse.jetty.io.ManagedSelector.createEndPoint(ManagedSelector.java:402)
	at org.eclipse.jetty.io.ManagedSelector.access$2100(ManagedSelector.java:65)
	at org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run(ManagedSelector.java:1069)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
```

No bytes are sent.

Reply:
Fixed: Alternate behavior for server side failure latch based on the existence of client side protocols on runtime JVM.